### DISCUSSION TITLE
Defining the specified indices for subdomains and boundaries

### DISCUSSION THREAD

**Discussion by EomAA** [firedrakeproject/firedrake#2677] [Firedrake support]
Hi folks,

I am transitioning and moving on to the Firedrake from FEniCS because it seems to me more flexible and easier to implement a parallelization of large-scale problems. 

In the FEniCS, I usually used a couple of functions 

> `"MeshFunction", "mark", and "CompiledSubDomain"`

 in order to define the specific indices for subdomains and boundaries based on the coordinates things. 

Looking at the webpage of Firedrake, it seems that only way to do this is to use the outsourcing mesh generator, like the Gmsh.

The same functions are supported in Firedrake or is there a something similar feature?

I look forward to hearing from you all and thank you in advance!!

**Comment by dham** [comment]
Basically you need to get the mesh generator to do this. You can work around it sometimes using indicator functions, but there's no straightforward way to change the markings on the mesh.

**Reply by EomAA** [reply]
Thanks a lot for the clarification! 
Using the mesh generator like the Gmsh is not trivial for me, but I can try it. 

Just in case, could you please recommend any resources for the indicator functions you mentioned?

**Reply by dham** [reply]
By indicator functions, I simply mean that you do something like:
```python3
dg0 = FunctionSpace(mesh, "DG", 0)
x = SpatialCoordinate(mesh)
indicator = interpolate(conditional(gt(x[0], 0.5), 1., 0.), dg0)
```
and then multiply your integrand by `indicator` or `1-indicator` to switch it on and off in appropriate parts of the domain. This is crude but effective.

**Comment by ksagiyam** [comment]
This PR will facilitate (to some extent) marking entities using indicator functions: https://github.com/firedrakeproject/firedrake/pull/2682

Specifically, you might be interested in this test: https://github.com/firedrakeproject/firedrake/pull/2682/commits/d7f16e21eba6b65e0d24bba9ca664551885d7b87
